source/lac/sparsity_tools.cc

```diff
@@ -39,15 +39,275 @@ extern "C"
+      *ierr = ZOLTAN_OK;
+
+      Assert( globalID != nullptr , ExcInternalError() );
+      Assert( localID != nullptr , ExcInternalError() );
+
+      //set global degrees of freedom
+      auto n_dofs = graph->n_rows();
+      for (int i=0; i < n_dofs; i++)
+        {
+          globalID[i] = i;    //global ids run from 0 to n_dofs-1
+          localID[i] = i;     //same as global ids
+        }
+
+      return;
+    }
+#endif
+
+//Query function unused if zoltan is not installed
+#ifdef DEAL_II_TRILINOS_WITH_ZOLTAN
+    void get_num_edges_list(void *data, int sizeGID, int sizeLID,
+                            int num_obj,
+                            ZOLTAN_ID_PTR globalID, ZOLTAN_ID_PTR localID,
+                            int *numEdges, int *ierr)
+    {
+      SparsityPattern *graph = reinterpret_cast<SparsityPattern *>(data);
+
+      *ierr = ZOLTAN_OK;
+
+
+      Assert ( numEdges != nullptr , ExcInternalError() );
+
+      for (int i=0; i<num_obj; ++i)
+        numEdges[i] = graph->row_length(globalID[i])-1;
+      return;
+    }
+#endif
+
+//Query function unused if zoltan is not installed
+#ifdef DEAL_II_TRILINOS_WITH_ZOLTAN
+    void get_edge_list(void *data, int sizeGID, int sizeLID,
+                       int num_obj, ZOLTAN_ID_PTR globalID, ZOLTAN_ID_PTR localID,
+                       int *num_edges,
+                       ZOLTAN_ID_PTR nborGID, int *nborProc,
+                       int wgt_dim, float *ewgts, int *ierr)
+    {
+      SparsityPattern *graph = reinterpret_cast<SparsityPattern *>(data);
+      *ierr = ZOLTAN_OK;
+
+      ZOLTAN_ID_PTR nextNborGID = nborGID;
+      int *nextNborProc = nborProc;
+
+      Assert( nextNborGID != nullptr , ExcInternalError() );
+      Assert( nextNborProc != nullptr , ExcInternalError() );
+
+      //Loop through rows corresponding to indices in globalID implicitly
+      for (int i=0; i < num_obj; ++i)
+        {
+          //Loop through each column to find neighbours
+          for ( SparsityPattern::iterator col=graph->begin(i);
+                col < graph->end(i); ++col )
+            //Ignore diagonal entries. Not needed for partitioning.
+            if ( i != col->column() )
+              {
+                *nextNborGID++ = col->column();
+                *nextNborProc++ = 0;    //All the vertices on processor 0
+              }
+        }
+      return;
+    }
+#endif
+
+
+    void partition_zoltan (const SparsityPattern     &sparsity_pattern,
+                           const unsigned int         n_partitions,
+                           std::vector<unsigned int> &partition_indices)
+    {
+      // Make sure that ZOLTAN is actually
+      // installed and detected
+#ifndef DEAL_II_TRILINOS_WITH_ZOLTAN
+      (void)sparsity_pattern;
+      AssertThrow (false, ExcInternalError());
+#else
+
+      //MPI environment must have been initialized by this point.
+      std::unique_ptr<Zoltan> zz = std::make_unique<Zoltan>(MPI_COMM_SELF);
```

Comment:
please use `std_cxx14::make_unique`, this is itnernal namespace in deal.II which will be equal to `std::` if `c++14` is there or will fallback to `boost` otherwise. That shall fix compiler errors on the tester.